src/libs/actions/IOU.js

```diff
@@ -201,41 +237,113 @@ function createIOUSplitGroup(params) {
+ * @param {Object} moneyRequestAction - the create IOU action we are cancelling
  */
-function rejectTransaction({
-    reportID, chatReportID, transactionID, comment,
-}) {
-    Onyx.merge(ONYXKEYS.TRANSACTIONS_BEING_REJECTED, {
-        [transactionID]: true,
-    });
-    DeprecatedAPI.RejectTransaction({
-        reportID,
+function cancelMoneyRequest(chatReportID, iouReportID, type, moneyRequestAction) {
+    const chatReport = chatReports[`${ONYXKEYS.COLLECTION.REPORT}${chatReportID}`];
+    const iouReport = iouReports[`${ONYXKEYS.COLLECTION.REPORT_IOUS}${iouReportID}`];
+    console.log("iouReport ", iouReport);
+    const transactionID = moneyRequestAction.originalMessage.IOUTransactionID;
+
+    // Get the amount we are cancelling
+    const amount = moneyRequestAction.originalMessage.amount;
+
+    const newSequenceNumber = Report.getMaxSequenceNumber(chatReport.reportID) + 1;
+    chatReport.maxSequenceNumber = newSequenceNumber;
+    const optimisticReportAction = ReportUtils.buildOptimisticIOUReportAction(
+        newSequenceNumber,
+        type,
+        amount,
+        moneyRequestAction.originalMessage.currency,
+        moneyRequestAction.originalMessage.comment,
+        '',
         transactionID,
-        comment,
-    })
-        .then((response) => {
-            if (response.jsonCode !== 200) {
-                Log.hmmm('Error rejecting transaction', {error: response.error});
-                return;
-            }
+        iouReportID,
+        '',
+    );
 
-            const chatReport = response.reports[chatReportID];
-            const iouReport = response.reports[reportID];
-            Report.syncChatAndIOUReports(chatReport, iouReport);
-        })
-        .finally(() => {
-            // Setting as null deletes the transactionID
-            Onyx.merge(ONYXKEYS.TRANSACTIONS_BEING_REJECTED, {
-                [transactionID]: null,
-            });
-        });
+    const currentUserEmail = optimisticReportAction.actorEmail;
+
+    // The owner of the IOU report is the account who is owed money,
+    // hence if current user is the owner, cancelling a request will lower the total and vice versa for declining
+    // or if the current user is not an owner of the report.
+    if (currentUserEmail === iouReport.ownerEmail) {
+        iouReport.total += type === CONST.IOU.REPORT_ACTION_TYPE.CANCEL ? -amount : amount;
+    } else {
+        iouReport.total += type === CONST.IOU.REPORT_ACTION_TYPE.CANCEL ? amount : -amount;
+    }
+
+    // If we are cancelling the
+    if (iouReport.total === 0) {
+        chatReport.hasOutstandingIOU = false;
+    } else if (iouReport.total < 0) {
+        // The total sign has changed and hence we need to flip the manager and owner of the report.
+        const oldOwnerEmail = iouReport.ownerEmail;
+        iouReport.ownerEmail = iouReport.managerEmail;
+        iouReport.managerEmail = oldOwnerEmail;
+        iouReport.total = -iouReport.total;
+    }
```

Comment:
@luacmartins We will have to do this for Splits as well, there might be already existing request between the user A and some of the participant in which case the participant has requested less than is now being requested by user A so we will have to flip the sign and the manager and owner on that particular report.

Reply:
Thanks! Created a method in my [PR](https://github.com/Expensify/App/blob/8f84ef50307c054127d1b86cd3714e9d4ba08522/src/libs/IOUUtils.js#L28-L58)